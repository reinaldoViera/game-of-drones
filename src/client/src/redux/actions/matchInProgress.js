import gql from 'graphql-tag';

import {
  START_MATCH, SERVER_LOADING, ERROR_SERVER_LOADING, ADD_PLAYER_1, ADD_PLAYER_2, RESET_GAME
} from "../constants/matchInProgress";
import {client} from './common'

export const resetMatch = (payload) => ({
  type: RESET_GAME,
  payload
})

export const fetchStart = () => ({
  type: SERVER_LOADING,
  payload: true
})
export const fetchEnd = () => ({
  type: SERVER_LOADING,
  payload: false
})
export const fetchError = (error = true) => ({
  type: ERROR_SERVER_LOADING,
  payload: error
})

// Thunk
export const createStartMatch = (dispatch) => (p1Id, p2Id, configId, cb) => () => {
  const mutation = gql`
  mutation startMatch {
      startMatch(player1: ${p1Id}, player2: ${p2Id}, config: ${configId}) {
        id
      }
  }
  `;
  dispatch(fetchStart());
  client.mutate({ mutation }).then(({ data })=> {
    const match = data.startMatch;
    dispatch(setStartMatch(match.id, p1Id, p2Id, configId, false))
    typeof cb === 'function' && cb(match);
    dispatch(fetchEnd());
  }).catch( e => dispatch(fetchError(e)))
};


export const setStartMatch = (matchId, player1, player2, configurationId, loading = true) => ({
  type: START_MATCH,
  payload: {
    matchId,
    winner: undefined,
    configurationId,
    loading,
    player1,
    player2,
  }
})

export const setPlayer1 = (id) => ({
  type: ADD_PLAYER_1,
  payload: id
})

export const setPlayer2 = (id) => ({
  type: ADD_PLAYER_2,
  payload: id
})
