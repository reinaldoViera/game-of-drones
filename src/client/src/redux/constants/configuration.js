export const ADD_CONFIG = 'Add new config';
export const ADD_CONFIGS = 'Add configs';
export const UPD_CONFIG = 'Update a config';
export const SERVER_LOADING = 'Loading configs';
export const ERROR_SERVER_LOADING = 'Error loading configs';