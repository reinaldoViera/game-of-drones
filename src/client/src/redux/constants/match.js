export const ADD_MATCH = 'Add new match';
export const ADD_MATCHS = 'Add matchs';
export const UPD_MATCH = 'Update a match';
export const ERROR_SERVER_LOADING = 'Loading matchs';
export const SERVER_LOADING = 'Error Loading matchs';