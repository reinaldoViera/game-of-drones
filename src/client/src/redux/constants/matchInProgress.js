export const START_MATCH = 'Start new match'
export const SERVER_START_MATCH = 'Start new match in server'
export const SERVER_LOADING = 'Loading data from server'
export const ERROR_SERVER_LOADING = 'Error loading data from server'
export const ADD_PLAYER_1 = 'Add player 1'
export const ADD_PLAYER_2 = 'Add player 2'
export const RESET_GAME = 'Reset the match'