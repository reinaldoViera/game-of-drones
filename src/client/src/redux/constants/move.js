export const ADD_MOVE = 'Add new move';
export const ADD_MOVES = 'Add moves';
export const UPD_MOVE = 'Update a move';
export const SERVER_LOADING = 'Loading moves';
export const ERROR_SERVER_LOADING = 'Error loading moves';