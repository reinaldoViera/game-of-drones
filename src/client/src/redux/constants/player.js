export const ADD_PLAYER = 'Add new player';
export const ADD_PLAYERS = 'Add players';
export const UPD_PLAYER = 'Update a player';
export const SERVER_LOADING = 'Loading players';
export const ERROR_SERVER_LOADING = 'Error loading players';