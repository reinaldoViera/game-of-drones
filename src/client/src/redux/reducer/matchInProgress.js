import {
    START_MATCH,
    SERVER_LOADING,
    ERROR_SERVER_LOADING,
    ADD_PLAYER_1,
    ADD_PLAYER_2,
    RESET_GAME
} from "../constants/matchInProgress";

const initialState = {
    matchId: undefined,
    rounds: [],
    loading: false,
    configurationId: '1',
    error: false,
    winner: undefined,
    player1: undefined,
    player2: undefined,
}

export default (state = initialState, {
    type,
    payload
}) => {
    switch (type) {

        case START_MATCH:
            return { ...payload
            }

        case SERVER_LOADING:
            return {
                ...state,
                error: payload ? false : state.error,
                loading: payload
            }
        case ERROR_SERVER_LOADING:
            return {
                ...state,
                error: payload
            }
        case ADD_PLAYER_1:
            return {
                ...state,
                player1: payload
            }
        case ADD_PLAYER_2:
            return {
                ...state,
                player2: payload
            }
        case RESET_GAME:
            return initialState;
        default:
            return state
    }
}