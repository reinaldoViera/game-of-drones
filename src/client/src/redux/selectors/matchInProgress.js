import { createSelector } from 'reselect'

export const getPlayer1 = createSelector([
    (state) => state.matchInProgress.player1,
    (state) => state.player.players,
], (player1, players) => {

    return player1 ? players[player1] : false;
})
export const getPlayer2 = createSelector([
    (state) => state.matchInProgress.player2,
    (state) => state.player.players,
], (player2, players) => {

    return player2 ? players[player2] : false;
})

export const getCurrentMoves = createSelector([
    (state) => state.matchInProgress.configurationId,
    (state) => state.configuration.configurations,
    (state) => state.move.moves,
], (configurationId, configs = {}, moves = {}) => {
    const currentConfig = configs[configurationId] || {};
    const currentMoves = currentConfig.move_types || [];
    return currentMoves.map(m => moves[m.id])
})