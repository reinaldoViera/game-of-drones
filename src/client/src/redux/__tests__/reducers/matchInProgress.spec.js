import matchInProgressReducer from '../../reducer/matchInProgress'
import * as actions from '../../constants/matchInProgress'

describe('Reducer for Match In progress', () => {
    test('should give inital state', () => {
        const state = matchInProgressReducer(undefined, {
            type: 'UNHANDLED'
        });
        expect(state).toMatchSnapshot()
    });
    test('should add Players', () => {
        let state = matchInProgressReducer(undefined, {
            type: actions.ADD_PLAYER_1,
            payload: 1
        });
        expect(state).toMatchSnapshot()
        state = matchInProgressReducer(undefined, {
            type: actions.ADD_PLAYER_2,
            payload: 2
        });
        expect(state).toMatchSnapshot()
    })

})