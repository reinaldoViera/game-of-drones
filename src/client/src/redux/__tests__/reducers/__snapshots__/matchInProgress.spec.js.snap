// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer for Match In progress should add Players 1`] = `
Object {
  "configurationId": "1",
  "error": false,
  "loading": false,
  "matchId": undefined,
  "player1": 1,
  "player2": undefined,
  "rounds": Array [],
  "winner": undefined,
}
`;

exports[`Reducer for Match In progress should add Players 2`] = `
Object {
  "configurationId": "1",
  "error": false,
  "loading": false,
  "matchId": undefined,
  "player1": undefined,
  "player2": 2,
  "rounds": Array [],
  "winner": undefined,
}
`;

exports[`Reducer for Match In progress should give inital state 1`] = `
Object {
  "configurationId": "1",
  "error": false,
  "loading": false,
  "matchId": undefined,
  "player1": undefined,
  "player2": undefined,
  "rounds": Array [],
  "winner": undefined,
}
`;
